### ISSUE TITLE
Incomplete instructions about building the compiler

### ISSUE THREAD

**Issue by goldfirere** [ocaml/ocaml#11499]
I'm just getting started in hacking on OCaml, so I thought I'd try building the compiler. I ran into a few gotchas in building. I'm reporting them here so that they might be improved for future contributors.

- `make` alone is not enough in order to use the built `ocamlc`, which tries to access `ocamlrun` in an absolute location. Instead, I've been advised that I have to install before trying out the compiler.

- `make tests`, I've been told, is the slow way to run the testsuite; `cd testsuite; make parallel` is the fast way.

**Comment by nojb** [comment]
> * which tries to access `ocamlrun` in an absolute location.

The non-relocatability of the toolchain is a long-standing issue; but I believe that there may be improvements coming soon on this (cc @dra27). For my part, I typically do one of: 1) `./configure --prefix $(pwd)/local` and then do `make; make install` to install the compiler in a subdirectory of the repository, or 2) leave it uninstalled and then run the binaries by passing them explicitly to the runtime: `runtime/ocamlrun ocamlc -I stdlib <more args>`.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2022/08/19 08:29 -0700):
> > * which tries to access `ocamlrun` in an absolute location.
>
> The non-relocatability of the toolchain is a long-standing issue; but I believe that there may be improvements coming soon on this (cc @dra27). For my part, I typically do one of: 1) `./configure --prefix $(pwd)/local` and then do `make; make install` to install the compiler in a subdirectory of the repository, or 2) leave it uninstalled and then run the binaries by passing them explicitly to the runtime: `runtime/ocamlrun ocamlc -I stdlib <more args>`.
>
`-nostdlib` may also be handy here otherwise the system will look for
libraries in its install directory, which may be either not found, or
wrong, corresponding to another compiler install.

**Comment by goldfirere** [comment]
Yes, I did something similar in the end, installing to a directory local to the repo. Not hard to do -- but it would be great if the instructions had this pointer. Thanks for the quick responses!

**Comment by shindere** [comment]
Richard Eisenberg (2022/08/19 10:16 -0700):
> Yes, I did something similar in the end, installing to a directory
> local to the repo. Not hard to do -- but it would be great if the
> instructions had this pointer. Thanks for the quick responses!

Would you feel like submitting a PR with the update to the documentation
that you feel would have been helpful ot you?

**Comment by goldfirere** [comment]
Happy to do so -- but I wanted to make sure that installing in a subdirectory is a suitable fix. It sounds like it is, so I'll prepare the PR.

**Comment by shindere** [comment]
Richard Eisenberg (2022/08/19 11:10 -0700):
> Happy to do so -- but I wanted to make sure that installing in a
> subdirectory is a suitable fix. It sounds like it is, so I'll prepare
> the PR.

Cool! Will be happy to review.

**Comment by gasche** [comment]
Happy and honored to see you around, @firere.

- You never need to install the compiler, but `./ocamlc` does not work without installing, instead you use, uh, something like `./runtime/ocamlrun ocamlc -nostdlib -I stdlib`. I think this is better than trying to install, because in iterative development cycles there are times where the distribution as a whole is not coherent (you may not want to install it) but `ocamlrun` and `ocamlc` are -- so that works. (Typically: hack on the compiler, rebuild `ocamlc` but not `ocamlopt`, and then quickly test something.)

- `make parallel` is mentioned in [HACKING.adoc : Useful Makefile targets](https://github.com/ocaml/ocaml/blob/trunk/HACKING.adoc#useful-makefile-targets) and in [testsuite/HACKING.adoc](https://github.com/ocaml/ocaml/blob/trunk/testsuite/HACKING.adoc). It does not work on all systems (it tries to work with GNU parallel or BSD parallel, I think?), so we cannot advertise it as the default command.

**Comment by goldfirere** [comment]
Thanks, @gasche! Nice to see you again, as well. :)

What you say makes sense -- and indeed matches my limited experience. That is, when I edited the compiler and then ran `make; make install`, running `./ocamlc` witnessed my changes, but `./output/bin/ocamlc` did not. This was surprising, but not worth really digging into. Perhaps this is why I created an Issue, not a PR, because I wasn't sure what the best practice was.

And thanks for pointing me down-document about the testsuite. I will spend more time reading it!

**Comment by gasche** [comment]
I believe there is a documentation issue, in particular the `-nostdlib -I stdlib` dance is not documented. It is mentioned in [runtime/HACKING.adoc](https://github.com/ocaml/ocaml/blob/trunk/runtime/HACKING.adoc) but assumed to be known already. So I guess a new section of our root HACKING.adoc is in order; "Running the built compilers"?

**Comment by shindere** [comment]
Gabriel Scherer (2022/08/19 21:24 -0700):
> I believe there is a documentation issue, in particular the `-nostdlib
> -I stdlib` dance is not documented. It is mentioned in
> [runtime/HACKING.adoc](https://github.com/ocaml/ocaml/blob/trunk/runtime/HACKING.adoc) but assumed to be known already. So I guess a new section of our root HACKING.adoc is in order; "Running the built compilers"?

How about "Running the compilers from sources"?

**Comment by kayceesrk** [comment]
For someone new to the OCaml compiler who is building the compiler from sources, the path that they may take is `README.adoc -> INSTALL.adoc` to [this section](https://github.com/ocaml/ocaml/blob/trunk/INSTALL.adoc#building-the-compiler). Perhaps adding a note here about running the compiler built from sources here or a link to the relevant section in a different document that describes it would be useful.

I often do 

```bash
$ ./configure --prefix=$(pwd)/_install
$ make -j
$ make install
$ ./_install/bin/ocamlopt
```

It may be useful to recommend the `--prefix` argument for someone building the compiler from the sources so that they can actually run the compiler. 

We may also consider recommending `make parallel` in the ["(Optional) Running the testsuite" section](https://github.com/ocaml/ocaml/blob/trunk/INSTALL.adoc#optional-running-the-testsuite).

**Comment by shindere** [comment]
Yeah, `--preifx` sounds sensible, for those who don't mind installing
somewhere.

Since it adds a stepp to the dev cycle, though, we could also imagine
providing scripts for runnning compilers from sources, perhaps.

**Comment by lpw25** [comment]
I think that [this section](https://github.com/ocaml/ocaml/blob/trunk/HACKING.adoc#your-first-compiler-modification) is the one @goldfirere used and it has a couple of short-comings:
1. It says:

    > 3\. Try the newly built compiler binaries ocamlc, ocamlopt or their .opt version. [...]

    having just built the compiler without a prefix and only having run `make`. This doesn't actually work if you do the obvious thing, and the text doesn't even tell you the obvious way to do it.
    
2. It says:

    > 5\. Run the testsuite from time to time.
    >
    >     make tests
    
    which is bad advice since you should probably do `make -C testsuite parallel`. Could we perhaps add a `tests-parallel` target to the main Makefile that does the same as `make tests` but with parallel? Then we could recommend that here.

**Comment by shindere** [comment]
Leo White (2022/08/20 02:46 -0700):
>     which is bad advice since you should probably do `make -C
>     testsuite parallel`. Could we perhaps add a `tests-parallel`
>     target to the main Makefile that does the same as `make tests` but
>     with parallel? Then we could recommend that here.

Wouldn't it be better, say if ocamltest is enabled, that configure
checks for `parallel` and then chooses the appropriate default target?

**Comment by dra27** [comment]
The relocatable compiler patches won't ever help with running bytecode executables in the build tree (you'll always have to run them with `ocamlrun` manually). They would be able to help with the `-nostdlib -I stdlib` dance _if_ the compiler build system is switched to put artefacts in an installation-like tree. i.e. if the compiler's build system puts the stdlib in, say, `_build/lib/ocaml` and `ocamlopt.opt` in `_build/bin/` then `_build/bin/ocamlopt.opt` will _always_ use `_build/lib/ocaml/stdlib.cmxa` etc. regardless of what's in the installation prefix's libdir.

**Comment by dra27** [comment]
> What you say makes sense -- and indeed matches my limited experience. That is, when I edited the compiler and then ran `make; make install`, running `./ocamlc` witnessed my changes, but `./output/bin/ocamlc` did not. This was surprising, but not worth really digging into. Perhaps this is why I created an Issue, not a PR, because I wasn't sure what the best practice was.

It's just possible that `ocamlc.opt` didn't get rebuilt in this case - when installed `ocamlc` is a symlink to `ocamlc.opt`. The `ocamlc` in the root of the build tree is the bytecode version which gets installed as `ocamlc.byte`.

**Comment by shindere** [comment]
David Allsopp (2022/08/24 02:22 -0700):
> > What you say makes sense -- and indeed matches my limited experience. That is, when I edited the compiler and then ran `make; make install`, running `./ocamlc` witnessed my changes, but `./output/bin/ocamlc` did not. This was surprising, but not worth really digging into. Perhaps this is why I created an Issue, not a PR, because I wasn't sure what the best practice was.
>
> It's just possible that `ocamlc.opt` didn't get rebuilt in this case -
> when installed `ocamlc` is a symlink to `ocamlc.opt`. The `ocamlc` in
> the root of the build tree is the bytecode version which gets
> installed as `ocamlc.byte`.

That's another aspect I'd like to change: that the programs do not get
renamed when they are installed.

**Comment by shindere** [comment]
Finally, I am unsure about what we shhould do regarding this issue.

Is there something we want to improve regarding the documentation and,
if so, is there a consensus about what and where it should go?

Or is it rather a matter of changing the build system?

Or perhaps, more likely, both? In whihc case perhaps we should break
this issue, maybe address one part (the doc) directly and open another,
more build-system specific issue?

**Comment by shindere** [comment]
Leo White (2022/08/20 02:46 -0700):
> I think that [this section](https://github.com/ocaml/ocaml/blob/trunk/HACKING.adoc#your-first-compiler-modification) is the one @goldfirere used and it has a couple of short-comings:
> 1. It says:
>
>     > 3\. Try the newly built compiler binaries ocamlc, ocamlopt or their .opt version. [...]
>
>     having just built the compiler without a prefix and only having
>     run `make`. This doesn't actually work if you do the obvious
>     thing, and the text doesn't even tell you the obvious way to do
>     it.

See #11518 for a proposal for an attempt
 to address this.

> 2. It says:
>
>     > 5\. Run the testsuite from time to time.
>     >
>     >     make tests
>
>     which is bad advice since you should probably do `make -C
>     testsuite parallel`. Could we perhaps add a `tests-parallel`
>     target to the main Makefile that does the same as `make tests` but
>     with parallel? Then we could recommend that here.

I think I already proposed this, but let me say it agian, just in case
and perhaps in a more precise way.

The suggestion is that, if `ocamltest is enabled`, then `configure`
could look for `parallel` and make sure it is used by default if
present, by the `tests` target. If `parallel` is not found, `configure`
could issue a warning and let the tests be run sequentially.

Would that be okay?

If yes I can work on it but won't start before I'm sure it will be fine.

**Comment by shindere** [comment]
David Allsopp (2022/08/24 02:20 -0700):
> The relocatable compiler patches won't ever help with running bytecode
> executables in the build tree (you'll always have to run them with
> `ocamlrun` manually). They would be able to help with the `-nostdlib
> -I stdlib` dance _if_ the compiler build system is switched to put
> artefacts in an installation-like tree. i.e. if the compiler's build
> system puts the stdlib in, say, `_build/lib/ocaml` and `ocamlopt.opt`
> in `_build/bin/` then `_build/bin/ocamlopt.opt` will _always_ use
> `_build/lib/ocaml/stdlib.cmxa` etc. regardless of what's in the
> installation prefix's libdir.

HAving a build tree that has the similar structure than the install tree
looks as a desirable goal to me. I hope we will reach it. It would also
help with testing the compiler. At the moment, ocamltest can only test a
compiler in its source/build tree, which are currently not distinguished
from each other. It is of course possible to extend ocamltest so that it
can test compilers both before and after they are installed, but the
principel itself of such an extension soudns bad to me.
